### ISSUE TITLE
Training on Multiclass

### ISSUE THREAD

**Issue by DrewNF** [Russell91/TensorBox#24]
Hi! I'd like to know if possible, how to modify/apply tensorbox to multiclass.
I'm working on it and I'd like to know if someone is working on to!
Here my GitHub: https://github.com/DrewNF/Tensorflow_VID_Video_Object_Tracking

**Comment by DrewNF** [comment]
Hi! sorry if I make you pissing off, but i'm really interest to adapt you work for multiclass and with your support i could be a great contributor, we can work with the hypes parameter "num_classes" ?? Inception that you hack to use, yet support multiclass learning, we can use this parameter and led the OverFeat to learn more label shapes?

**Comment by Russell91** [comment]
Yes, it is definitely possible to support multiclass. The code is set up so that this should be a relatively small change, but it just hasn't been implemented yet. The first thing would be to set the silhouetteID of each rectangle to be it's class in the idl file,

    from utils.annolist import AnnotationLib as al
    r = al.AnnoRect()
    r.silhouetteID = 2

Once you have the idl file, you can parse that in train_utils.py and load in values of 1,2,3,... into the box_flags. The code already considers any value > 0 to be an object for regression purposes. So you ignore the effects on regression and instead just use a softmax to guess the class of each box.

**Comment by DrewNF** [comment]
Hi Russell,
After resolving some problems in my algorithm and start train TensorBox, I start to modify the code for the multiclass training.
After some reverse eng I understand this:
In train.py is called train_utils.load_data_gen that call load_idl_tf that finally call data_utils.annotation_to_h5, in this traceback is the load of the file so also of the flags to set from the idl file.
I don't understand how are assigned and readed from the idl file the informations in the data_utils.annotation_to_h5 function:
here initialized:
 box_flags = np.zeros((1, cells_per_image, 1, max_len, 1), dtype = np.float)
here setted all to 1:
box_flags[0, cidx, 0, 0:len(unsorted_boxes), 0] = 1
but each boxe in the boxes have different label.
Bests Regards.

**Comment by DrewNF** [comment]
Hi Russel! 
I implement the multiclass, and take some hints from your develop branch for the question above.
Now I'm training it on the VID dataset,I will provide you a weight file,if you are interested.
You can see what I'm doing on my branch.

**Comment by DrewNF** [comment]
HI Russel! I'm testing the multiclass model, but the pre_confidence returned with the n°shape, always have the first element as the higher value (so theoretically the founded shape silhouette class), but really it found differents ones.
I get wrong something?

**Comment by DrewNF** [comment]
Hi Russell! I made many tests to figure out why doesn't work, so I wonna know if the idl pattern that I'm using is the right one or not:
"./Data/VID/train/ILSVRC2015_VID_train_0003/ILSVRC2015_train_00080007/000029.JPEG" @29 : (319.0,136.0,385.0,308.0) /13 :1;
"./Data/VID/train/ILSVRC2015_VID_train_0003/ILSVRC2015_train_00080007/000077.JPEG" @77 : (276.0,138.0,353.0,279.0) /13 :1;

Let me know because right know it raise this error:

  File "train.py", line 532, in <module>
    main()
  File "train.py", line 529, in main
    train(H, test_images=[])
  File "train.py", line 455, in train
    d = gen.next()
  File "/home/users/aferri/TensorBox/utils/train_utils.py", line 78, in load_data_gen
    for d in data:
  File "/home/users/aferri/TensorBox/utils/train_utils.py", line 31, in load_idl_tf
    annolist = al.parse(idlfile)
  File "/home/users/aferri/TensorBox/utils/annolist/AnnotationLib.py", line 710, in parse
    annolist = parseIDL(filename)
  File "/home/users/aferri/TensorBox/utils/annolist/AnnotationLib.py", line 806, in parseIDL
    rect.silhouetteID = int(rectSeg[posSlash+1:])
ValueError: invalid literal for int() with base 10: '13 :1'
Bests

**Comment by violivei** [comment]
I'm also pretty interested in multiclass classification. Any news about this feature?
How to set up the input file?

Thanks.

**Comment by DrewNF** [comment]
@violivei look the Issue: #42 , I also made a pull request: #50 where I add some referencies that can really help for structurize the .IDL file in the right way.

**Comment by DrewNF** [comment]
Hi russel! Maybe i'm on the right way to find why the label prediction doesn't work, when you load the box_flags their shape is (1, 300, 1, 1, 1) and seems you never load the right_numclasses in building the array and so it never really knows and predict the label because is only one class for him. I'm on the right way?

**Comment by Russell91** [comment]
Yea, that is correct! So you need to modify the size of those output prediction from 2-fold (the current model) to k-fold. In particular, the easiest way to do this would be after the rezoom layer.

**Comment by Zigngit** [comment]
Is there anything progress on this one ? I need this too.
